artiq/gateware/rtio/phy/urukul_monitor.py

```diff
@@ -0,0 +1,50 @@
+from migen import *
+
+from artiq.coredevice.spi2 import SPI_CONFIG_ADDR, SPI_DATA_ADDR, SPI_END
+from artiq.coredevice.urukul import CS_DDS_CH0, CS_DDS_MULTI
+
+
+class UrukulMonitor(Module):
+    def __init__(self, spi_rtlink, nchannels=4):
+        # [0:3] = register address, [4:7] = data_high, [8:11] = data_low (or just data for 32-bit transfers)
+        self.probes = Array([Signal(32) for i in range(3 * nchannels)])
+        self.rtlink = spi_rtlink
+
+        current_address = Signal.like(self.rtlink.o.address)
+        current_data = Signal.like(self.rtlink.o.data)
+        self.sync.rio += If(
+            self.rtlink.o.stb,
+            current_address.eq(self.rtlink.o.address),
+            current_data.eq(self.rtlink.o.data)
+        )
+
+        cs = Signal(8)
+        length = Signal(8)
+        end = Signal()
+        self.sync.rio += If(
+            self.rtlink.o.stb & (current_address == SPI_CONFIG_ADDR),
```

Comment:
rtlink address is only valid in the cycle at which ``stb`` is asserted. Here they are shifted by one clock cycle (with the pipeline delay on ``current_address``) and this looks incorrect.
Stefan's code is full of problems and generally not an example to follow.